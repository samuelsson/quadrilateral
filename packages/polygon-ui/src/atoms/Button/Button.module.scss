[data-theme="light"] {
  --Button-background-color: var(--color-blue-gray-800);
  --Button-color: var(--color-white);
}

[data-theme="dark"] {
  --Button-background-color: var(--color-blue-gray-100);
  --Button-color: var(--color-blue-gray-800);
}

.Button {
  height: 48px;
  border: none;
  padding: 0 2rem;
  border-radius: 5px;
  font-weight: bold;
  background-color: var(--Button-background-color);
  color: var(--Button-color);
  text-decoration: none;
  text-transform: uppercase;

  transition-property: background, border, color, box-shadow;
  transition-timing-function: linear;
  transition-duration: 0.2s;

  &:hover:not(:disabled),
  &:focus:not(:disabled),
  &:active:not(:disabled) {
    cursor: pointer;
  }
}
