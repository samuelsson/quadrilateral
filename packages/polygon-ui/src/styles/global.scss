@import './abstracts/colors';

[data-theme='light'] {
  --global-background-color: #{$color-white};
  --global-color: #{$color-blue-gray-800};
}

[data-theme='dark'] {
  --global-background-color: #{$color-blue-gray-800};
  --global-color: #{$color-blue-gray-100};
}

body {
  background-color: var(--global-background-color);
  color: var(--global-color);
  font-family: sans-serif;
}
