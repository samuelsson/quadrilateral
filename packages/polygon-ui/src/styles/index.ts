export { colors } from './abstracts';
export { variables } from './abstracts';

export { default as global } from './global';
