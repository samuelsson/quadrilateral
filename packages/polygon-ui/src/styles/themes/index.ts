export { default as light } from './light';
export { default as dark } from './dark';

export { default as Theme } from './types';
