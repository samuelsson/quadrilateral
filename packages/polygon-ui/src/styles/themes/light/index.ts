import base from './base';
import button from './button';
import input from './input';

export default {
  base,
  button,
  input,
};
