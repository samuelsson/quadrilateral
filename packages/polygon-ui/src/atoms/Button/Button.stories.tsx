import * as React from 'react';
import { storiesOf } from '@storybook/react';
import styled from 'styled-components';
import { boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import Button from './Button';

const ButtonContainer = styled.div`
  button:not(:last-of-type) {
    margin-right: 1rem;
  }
`;

const appearance = ['primary', 'secondary', 'success', 'danger'];

storiesOf('Button', module)
  .addDecorator(storyFn => <ButtonContainer>{storyFn()}</ButtonContainer>)
  .add(
    'Default',
    (): JSX.Element => (
      <>
        {appearance.map(button => (
          <Button
            key={button}
            name={button}
            appearance={button}
            onClick={action('button-click')}
            disabled={boolean('Disabled', false)}
            bare={boolean('Bare', false)}
          >
            {button.toUpperCase()}
          </Button>
        ))}
      </>
    )
  );
