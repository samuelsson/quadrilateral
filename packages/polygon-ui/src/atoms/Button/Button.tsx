import React from 'react';
import styled, { css, FlattenSimpleInterpolation } from 'styled-components';
import { darken } from 'polished';
import { colors, variables } from '../../styles';
import { Theme } from '../../styles/themes';

type Type = 'button' | 'submit' | 'reset';
type Appearance = 'primary' | 'secondary' | 'success' | 'danger';

interface ButtonProps {
  name: string;
  children: React.ReactNode;
  appearance?: Appearance;
  type?: Type;
  bare?: boolean;
  disabled?: boolean;
  theme?: Theme;
}

interface BaseStyle {
  foreground: string;
  background: string;
  bare?: boolean;
}

const baseStyle = ({
  foreground,
  background,
  bare,
}: BaseStyle): FlattenSimpleInterpolation => css`
  height: 48px;
  border: none;
  padding: 0 2rem;
  border-radius: ${variables.borderRadius};
  box-shadow: ${!bare && `0 2px 8px ${colors.gray[300]}`};
  background-color: ${bare ? 'transparent' : background};
  color: ${bare ? background : foreground};
  font-weight: bold;
  text-decoration: none;
  text-transform: uppercase;

  transition-property: background, border, color, box-shadow;
  transition-timing-function: linear;
  transition-duration: 0.2s;

  &:hover:not(:disabled),
  &:focus:not(:disabled),
  &:active:not(:disabled) {
    cursor: pointer;
    background-color: ${darken(0.15, background)};
    box-shadow: 0 2px 10px ${colors.gray[400]};
    color: ${foreground};
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.5;
    box-shadow: none;
  }
`;

const Button: React.FC<ButtonProps> = styled.button<ButtonProps>`
  ${({
    theme,
    appearance: buttonAppearance,
    bare,
  }): FlattenSimpleInterpolation => {
    const { white, red, green, blue } = colors;

    switch (buttonAppearance) {
      case 'secondary':
        return baseStyle({
          foreground: white,
          background: blue,
          bare: !!bare,
        });
      case 'success':
        return baseStyle({
          foreground: white,
          background: green,
          bare: !!bare,
        });
      case 'danger':
        return baseStyle({
          foreground: white,
          background: red,
          bare: !!bare,
        });
      default:
        return baseStyle({
          foreground: theme.button.primary.foreground,
          background: theme.button.primary.background,
          bare: !!bare,
        });
    }
  }};
`;

export default Button;
