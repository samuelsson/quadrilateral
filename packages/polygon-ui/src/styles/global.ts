import { createGlobalStyle } from 'styled-components';
import { Theme } from './themes';

type GlobalStyleProps = {
  theme: Theme;
};

export default createGlobalStyle<GlobalStyleProps>`
    body {
      background-color: ${({ theme }): string => theme.base.backgroundColor};
      color: ${({ theme }): string => theme.base.color};
      font-family: sans-serif;
    }
`;
