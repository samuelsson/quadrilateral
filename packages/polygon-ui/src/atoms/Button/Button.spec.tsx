import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, RenderResult } from '@testing-library/react';

import { Theme, dark, light } from '../../styles/themes';
import Button from './Button';

describe('Button', (): void => {
  const renderWithTheme = (element: JSX.Element, theme: Theme): RenderResult =>
    render(<ThemeProvider theme={theme}>{element}</ThemeProvider>);

  it('should match snapshot with light theme', (): void => {
    const { asFragment } = renderWithTheme(
      <Button name="button">Light</Button>,
      light
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot with dark theme', (): void => {
    const { asFragment } = renderWithTheme(
      <Button name="button">Dark</Button>,
      dark
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
