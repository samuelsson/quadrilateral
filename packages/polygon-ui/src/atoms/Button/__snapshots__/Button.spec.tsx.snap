// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot with dark theme 1`] = `
<DocumentFragment>
  <button
    class="sc-AxjAm glKCZb"
    name="button"
  >
    Dark
  </button>
</DocumentFragment>
`;

exports[`Button should match snapshot with light theme 1`] = `
<DocumentFragment>
  <button
    class="sc-AxjAm gTiQGs"
    name="button"
  >
    Light
  </button>
</DocumentFragment>
`;
